src/ConstructiveSolidGeometry/SurfacePrimitives/ConeMantle.jl

```diff
@@ -1,150 +1,346 @@
 end
 
-function ConeMantle(t::Torus{T}; θ = π/2) where {T}
-    r_tubeMin::T, r_tubeMax::T = get_r_tube_limits(t)
-    θ = T(mod(θ,2π))
-    sθ, cθ = sincos(θ)
-    if θ > T(0) && θ < T(π)
-        rbot = geom_round(t.r_torus + r_tubeMin*cθ)
-        rtop = geom_round(t.r_torus + r_tubeMax*cθ)
-        zMin = geom_round(t.z + r_tubeMin*sθ)
-        zMax = geom_round(t.z + r_tubeMax*sθ)
-    elseif θ > T(π) && θ < T(2π)
-        rtop = geom_round(t.r_torus + r_tubeMin*cθ)
-        rbot = geom_round(t.r_torus + r_tubeMax*cθ)
-        zMax = geom_round(t.z + r_tubeMin*sθ)
-        zMin = geom_round(t.z + r_tubeMax*sθ)
-    else
-        @error "Cone Mantle not defined for torroidal cordinate θ = 0 or θ = π. Use Annulus"
-    end
-    r = rbot == rtop ? T(rbot) : (T(rbot), T(rtop))
-    z = zMax == -zMin ? T(zMax) : T(zMin)..T(zMax)
-    ConeMantle( T, r, t.φ, z)
+radius_at_z(hZ::T, rBot::T, rTop::T, z::T) where {T} = iszero(hZ) ? rBot : rBot + (hZ+z)*(rTop - rBot)/(2hZ) 
+radius_at_z(cm::ConeMantle{T,T}, z::T) where {T} = cm.r
+radius_at_z(cm::ConeMantle{T,Tuple{T,T}}, z::T) where {T} = radius_at_z(cm.hZ, cm.r[1], cm.r[2], z)
+
+function normal(cm::ConeMantle{T,T}, pt::CartesianPoint{T}) where {T}
+    pto = _transform_into_object_coordinate_system(pt, cm)
+    cyl = CylindricalPoint(pto)
+    return CartesianVector(_transform_into_global_coordinate_system(
+            CartesianPoint(CylindricalPoint{T}(cyl.r, cyl.φ, zero(T))), cm))
+end
+function normal(cm::ConeMantle{T,Tuple{T,T},<:Any,:inwards}, pt::CartesianPoint{T}) where {T}
+    pto = _transform_into_object_coordinate_system(pt, cm)
+    cyl = CylindricalPoint(pto)
+    Δr = cm.r[2] - cm.r[1]
+    z = cyl.r * Δr / 2cm.hZ
+    return CartesianVector(_transform_into_global_coordinate_system(
+            CartesianPoint(CylindricalPoint{T}(-cyl.r, cyl.φ, z)), cm))
 end
+function normal(cm::ConeMantle{T,Tuple{T,T},<:Any,:outwards}, pt::CartesianPoint{T}) where {T}
+    pto = _transform_into_object_coordinate_system(pt, cm)
+    cyl = CylindricalPoint(pto)
+    Δr = -(cm.r[2] - cm.r[1])
+    z = cyl.r * Δr / 2cm.hZ
+    return CartesianVector(_transform_into_global_coordinate_system(
+            CartesianPoint(CylindricalPoint{T}( cyl.r, cyl.φ, z)), cm))
+end
+
+const FullConeMantle{T,D} = ConeMantle{T,Tuple{T,T},Nothing,D} # ugly name but works for now, should just be `ConeMantle`...
+const PartialConeMantle{T,D} = ConeMantle{T,Tuple{T,T},Tuple{T,T},D}
+
 
-function ConeMantle(;rbot = 1, rtop = 0, φMin = 0, φMax = 2π, zMin = -1/2, zMax = 1/2)
-    T = float(promote_type(typeof.((rbot, rtop, φMin, φMax, zMin, zMax))...))
-    r = rbot == rtop ? T(rbot) : (T(rbot), T(rtop))
-    φ = mod(T(φMax) - T(φMin), T(2π)) == 0 ? nothing : T(φMin)..T(φMax)
-    z = zMax == -zMin ? T(zMax) : T(zMin)..T(zMax)
-    ConeMantle(T, r, φ, z)
+extremum(cm::FullConeMantle{T}) where {T} = sqrt(cm.hZ^2 + max(cm.r[1], cm.r[2])^2)
+extremum(cm::PartialConeMantle{T}) where {T} = sqrt(cm.hZ^2 + max(cm.r[1], cm.r[2])^2)
```

Comment:
Again, we could change `max(cm.r[1], cm.r[2])` to `maximum(cm.r)`

Reply:
It probably does not really matter. 
I used `max` to avoid an iterator.